fuel-vm/src/interpreter/receipts.rs

```diff
@@ -0,0 +1,170 @@
+use core::ops::Index;
+use std::mem;
+
+use fuel_merkle::binary;
+use fuel_tx::Receipt;
+use fuel_types::{bytes::SerializableVec, Bytes32};
+
+#[derive(Debug, Default, Clone)]
+pub(crate) struct ReceiptsCtx {
+    receipts: Vec<Receipt>,
+    receipts_tree: binary::in_memory::MerkleTree,
+}
+
+impl ReceiptsCtx {
+    pub fn push(&mut self, mut receipt: Receipt) {
+        self.receipts_tree.push(receipt.to_bytes().as_slice());
+        self.receipts.push(receipt)
+    }
+
+    pub fn clear(&mut self) {
+        self.receipts_tree.reset();
+        self.receipts.clear();
+    }
+
+    pub fn root(&self) -> Bytes32 {
+        self.receipts_tree.root().into()
+    }
+
+    /// Get a mutable lock on this context
+    pub fn lock(&mut self) -> ReceiptsCtxMut {
+        ReceiptsCtxMut::new(self)
+    }
+
+    /// Recalculates the Merkle root of the receipts from scratch. This should
+    /// only be used when the list of receipts has been mutated externally.
+    fn recalculate_root(&mut self) {
+        self.receipts_tree.reset();
+        let receipts = self.as_ref().clone();
```

Comment:
maybe we should add a todo here about removing clone once `to_bytes()` no longer requires `&mut self`